client/client_deprecated.go

```diff
@@ -11,13 +14,13 @@ import "net/http"
 // server is upgraded.
 // Deprecated: use NewClientWithOpts
 func NewClient(host string, version string, client *http.Client, httpHeaders map[string]string) (*Client, error) {
-	return NewClientWithOpts(WithHost(host), WithVersion(version), WithHTTPClient(client), WithHTTPHeaders(httpHeaders))
+	return nil, errors.New("deprecated: use 'client.NewClientWithOpts(WithHost(host), WithVersion(version), WithHTTPClient(client), WithHTTPHeaders(httpHeaders))'")
 }
```

Comment:
Can we just remove this function so that downstreams can detect the change on the compilation time?

Reply:
That's a good point; it won't be a compile error.

What I tried to achieve is to make it something that cannot be ignored by project using the old code, but giving them enough guidance to fix it. Perhaps as an alternative, we could;

- change the signature to be incompatible (causing a compile failure); removing all returns.
- change the function's documentation to provide the instructions

Something like:

```go
// NewClient initializes a new API client for the given host and API version.
// It uses the given http client as transport.
// It also initializes the custom http headers to add to each request.
//
// It won't send any version information if the version number is empty. It is
// highly recommended that you set a version or your client may break if the
// server is upgraded.
//
// This function is deprecated, and is no longer functional.
//
// If you're using this function, replace it with the code below to get
// the equivalent (non-deprecated) code to initialize a client with a
// fixed API version:
//
//    client.NewClientWithOpts(
//        WithHost(host),
//        WithVersion(version),
//        WithHTTPClient(client),
//        WithHTTPHeaders(httpHeaders),
//    )
//
// We recommend to enable API version negotiation, so that the client
// automatically negotiate the API version to use when connecting with the
// daemon. To enable API version negotiation, use the WithAPIVersionNegotiation()
// option instead of WithVersion(version):
//
//    client.NewClientWithOpts(
//        WithHost(host),
//        WithAPIVersionNegotiation(),
//        WithHTTPClient(client),
//        WithHTTPHeaders(httpHeaders),
//    )
//
// Deprecated: use NewClientWithOpts
func NewClient(host string, version string, client *http.Client, httpHeaders map[string]string) {}

// NewEnvClient initializes a new API client based on environment variables.
// See FromEnv for a list of support environment variables.
//
// This function is deprecated, and is no longer functional.
//
// If you're using this function, replace it with the code below to get
// the equivalent (non-deprecated) code:
//
//    client.NewClientWithOpts(FromEnv)
//
// We recommend to enable API version negotiation, so that the client
// automatically negotiate the API version to use when connecting with the
// daemon. To enable API version negotiation, add the WithAPIVersionNegotiation()
// option:
//
//    client.NewClientWithOpts(FromEnv, WithAPIVersionNegotiation())
//
// Deprecated: use NewClientWithOpts
func NewEnvClient() {}
```

Reply:
Maybe just comment out the code without breaking the prototype and add deprecation document there?

Reply:
Yeah, I thought that having the stub would allow people to more easily find it back (at least, my usual approach is to have my IDE link to the right file) and to still have docs at `pkg.go.dev`, but perhaps just `grep` could work of course